### PR TITLE
273 Test module for BaseDateTransformer in line with new test set up

### PR DESCRIPTION
**Pull request by davidhopkinson26**
- Adds Test modules for BaseDateTransformer and BaseDateTwoColumnTransformer in line with the new test set up
- created new mixin classes to handle new_column_name arguments and for cases with two columns (this required some changes to comparison.py)
- renamed new_col_name to new_column_name to be consistent across th apckage and allow use of a shared mixin to handle validation.
- changed the name of BaseDropOriginaltransformer to DropOriginalTransformer.  This affected quite a few additional files.

I had some trouble in this caused by the high degree of coupling introduced by conftest.  If there is a bug initialising one transformer it can cause virtually all tests in the package to fail.  I have created issue #285 to explain in more detail some proposed changes to explore.  This is not urgent but the current set up can make some problems hard to debug (always scary when a change seems to break hundreds of unrelated tests!)

On the plus size I went round the houses a bit on class structures and what to put in mixins vs base classes in developing this PR and the new test set up meant that I mostly didn't need to make changes to tests when doing so as tests were grouped by function rather than mimicking the class structure or testing implementation.  This was a pleasant experience and I think bodes well for future refactoring efforts going smoothly!

### CODE REVIEW

File: tests/base_tests.py [hunk 1]
```diff
@@ -214,29 +244,6 @@ def test_list_length_error(
         ):
             uninitialized_transformers[self.transformer_name](**args)
 
-    @pytest.mark.parametrize(
```
Comment (Chip2916) [inline tests/base_tests.py:217]: Q: Just wondering, does the TestInit class in 'test_BaseTwoColumnTransformer.py' need updating to inherit 'NewColumnNameInitMixintests' as per 'test_EqualityChecker.py' now this has been removed from here? Asking as TestInit in 'test_BaseTwoColumnTransformer' inherits from 'TwoColumnListInitTests' as per - 

https://github.com/lvgig/tubular/blob/e92f8dc95cb6016a4d730fd1dc6a9977a6b46e69/tests/base/test_BaseTwoColumnTransformer.py#L9

File: tests/conftest.py [hunk 2]
```diff
@@ -137,10 +137,6 @@ def minimal_attribute_dict():
         "BaseTransformer": {
             "columns": ["a"],
         },
-        "BaseTwoColumnTransformer": {
```
Comment (Chip2916) [inline tests/conftest.py:140]: Q: Why is this removed, are we not using this anymore?

File: tests/dates/test_BaseDateTransformer.py [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+        for col in ["date_col_1", "date_col_2"]:
+            df[col] = pd.to_datetime(df[col]).dt.date
+
+        for col in ["datetime_col_1", "datetime_col_2"]:
+            df[col] = pd.to_datetime(df[col])
+
+        present_types = (
+            {"datetime64", "date"} if datetime_col == 0 else {"date", "datetime64"}
+        )
+        msg = rf"Columns fed to datetime transformers should be \['datetime64', 'date'\] and have consistent types, but found {present_types}. Please use ToDatetimeTransformer to standardise"
```
Comment (Chip2916) [inline tests/dates/test_BaseDateTransformer.py:108]: N: Is this message a bit misleading? Just to check my understanding we want both columns to have a consistent type, but the 'should be \['datetime64', 'date'\]' suggests that we want one dt64 and one dt column?

File: tubular/base.py [hunk 4]
```diff
@@ -241,50 +241,7 @@ def columns_check(self, X: pd.DataFrame) -> None:
                 raise ValueError(f"{self.classname()}: variable " + c + " is not in X")
 
 
-class BaseTwoColumnTransformer(BaseTransformer):
```
Comment (Chip2916) [inline tubular/base.py:244]: Q: Is this still required here - https://github.com/lvgig/tubular/blob/e92f8dc95cb6016a4d730fd1dc6a9977a6b46e69/tubular/comparison.py#L9

Comment (Chip2916) [reply tubular/base.py:244]: Actually, that's the next file, and can see you've changed it :)

File: tubular/dates.py [hunk 5]
```diff
@@ -9,11 +9,12 @@
 import pandas as pd
 
 from tubular.base import BaseTransformer
+from tubular.mixins import DropOriginalMixin, NewColumnNameMixin, TwoColumnMixin
 
 
-class BaseDateTransformer(BaseTransformer):
+class BaseDateTransformer(DropOriginalMixin, BaseTransformer):
```
Comment (Chip2916) [inline tubular/dates.py:15]: Q: Should NewColumnNameMixin be specified here?

File: tubular/mixins.py [hunk 6]
```diff
@@ -60,6 +60,30 @@ def drop_original_column(
...
+            msg = f"{self.classname()}: new_column_name should be str"
+            raise TypeError(msg)
+
+        self.new_column_name = new_column_name
+
+
+class TwoColumnMixin:
+    """helper to validate columns when exactly two columns are required"""
+
+    def check_two_columns(self, columns: list[str] | str) -> None:
```
Comment (Chip2916) [inline tubular/mixins.py:77]: Q: Can two column input be just a single str? Would that not fail the if not (isinstance(columns, list)) test

File: tests/conftest.py (hunk 2 above)
Comment (davidhopkinson26) [reply tests/conftest.py:140]: yes have refactored the logic form this transformer into two mixin classes NewColumnNameMixin and TwoColumnMixin so no longer have this base class.

File: tests/dates/test_BaseDateTransformer.py (hunk 3 above)
Comment (davidhopkinson26) [reply tests/dates/test_BaseDateTransformer.py:108]: This is set up to read the allowed types argument passed to the method so this error message might say datetime64 or date in it and this expression will accept either.

File: tests/base_tests.py (hunk 1 above)
Comment (davidhopkinson26) [reply tests/base_tests.py:217]: yes it does as I moved the NewColumnNameMixin tests into the mixn tests out of TwoColumnListInitTests.  Looks like it does already inherit them to me? Might be missing something here.

Comment (Chip2916) [reply tests/base_tests.py:217]: Ahh okay, this was my first comment but can see from your comments below we are no longer using BaseTwoColumnTransformer

File: tests/dates/test_BaseDateTransformer.py (hunk 3 above)
Comment (Chip2916) [reply tests/dates/test_BaseDateTransformer.py:108]: Ahh okay thanks :)

File: tubular/mixins.py (hunk 6 above)
Comment (Chip2916) [reply tubular/mixins.py]: Thanks David

File: tubular/dates.py (hunk 5 above)
Comment (Chip2916) [reply tubular/dates.py]: Cool thanks :)

**Review by Chip2916 — APPROVED** [review]
Thanks David, happy to approve
